sqlglot/optimizer/simplify.py

```diff
@@ -530,6 +552,134 @@ def simplify_concat(expression):
     return new_args[0] if len(new_args) == 1 else concat_type(expressions=new_args)
 
 
+def simplify_datetrunc(expression):
+    """Remove redundant DATE_TRUNC functions"""
+    if isinstance(expression, exp.DateTrunc) and expression.unit.name.lower() == "day":
+        return expression.this
+    return expression
+
+
+DateRange = t.Tuple[datetime.date, datetime.date]
+
+
+def _datetrunc_range(date: datetime.date, unit: str) -> t.Optional[DateRange]:
+    """
+    Get the date range for a DATE_TRUNC equality comparison:
+
+    Example:
+        _datetrunc_range(date(2021-01-01), 'year') == (date(2021-01-01), date(2022-01-01))
+    Returns:
+        tuple of [min, max) or None if a value can never be equal to `date` for `unit`
+    """
+    floor = date_floor(date, unit)
+
+    if date != floor:
+        # This will always be False, except for NULL values.
+        return None
+
+    return floor, floor + interval(unit)
+
+
+def _datetrunc_eq_expression(left: exp.Expression, drange: DateRange) -> exp.Expression:
+    """Get the logical expression for a date range"""
+    return exp.and_(
+        exp.GTE(this=left.copy(), expression=date_literal(drange[0])),
```

Comment:
left <= ...

Reply:
We only define the dunder comparison methods on `Condition`. This can lead to attribute error if, for some reason, `left` is not a `Condition`.
Perhaps we should define the dunder methods on Expression?

Reply:
yea let's move that?

Reply:
done